azkaban-common/src/main/java/azkaban/storage/StorageCleaner.java

```diff
@@ -0,0 +1,156 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package azkaban.storage;
+
+import static azkaban.Constants.ConfigurationKeys.AZKABAN_STORAGE_ARTIFACT_MAX_RETENTION;
+import static com.google.common.base.Preconditions.checkArgument;
+
+import azkaban.db.DatabaseOperator;
+import azkaban.project.ProjectLoader;
+import azkaban.spi.Storage;
+import azkaban.utils.Props;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.log4j.Logger;
+
+@Singleton
+public class StorageCleaner {
+
+  // Delete records of all older versions
+  static final String SQL_DELETE_OLDER_VERSIONS = "DELETE FROM project_versions WHERE project_id=? AND version < ?";
+
+  // The query must sort the versions in reverse order for the cleanup operation to work correctly!
+  static final String SQL_FETCH_PVR = "SELECT version, resource_id FROM project_versions WHERE project_id=? ORDER BY version DESC";
+
+  private static final Logger log = Logger.getLogger(StorageCleaner.class);
+  private final ProjectLoader projectLoader;
+  private final DatabaseOperator databaseOperator;
+  private final int maxArtifactsPerProject;
+  private final Storage storage;
+
+  @Inject
+  public StorageCleaner(final Props props, final Storage storage,
+      final DatabaseOperator databaseOperator, final ProjectLoader projectLoader) {
+    this.storage = storage;
+    this.databaseOperator = databaseOperator;
+
+    this.maxArtifactsPerProject = props.getInt(AZKABAN_STORAGE_ARTIFACT_MAX_RETENTION, 0);
+    this.projectLoader = projectLoader;
+    checkArgument(this.maxArtifactsPerProject >= 0,
+        String.format("Invalid value for %s : %d", AZKABAN_STORAGE_ARTIFACT_MAX_RETENTION,
+            this.maxArtifactsPerProject));
+
+    if (isCleanupPermitted()) {
+      log.info(String.format("%s Config: Max %d artifact(s) retained per project",
+          AZKABAN_STORAGE_ARTIFACT_MAX_RETENTION, this.maxArtifactsPerProject));
+    } else {
+      log.warn("Project cleanup disabled. All artifacts will be stored.");
+    }
+  }
+
+  @VisibleForTesting
+  boolean isCleanupPermitted() {
+    return this.maxArtifactsPerProject > 0;
+  }
+
+  /**
+   * Remove all but last N artifacts as configured by AZKABAN_STORAGE_ARTIFACT_MAX_RETENTION
+   *
+   * Since multiple versions can share the same filename, the algo is to collect all filenames and
+   * from them, remove the latest ones. The remaining ones are deleted by the respective storage.
+   *
+   * @param projectId project ID
+   */
+  public void cleanupProjectArtifacts(final int projectId) {
+    if (!isCleanupPermitted()) {
+      return;
+    }
+    final List<VersionResourcePair> pvrList = fetchPvrList(projectId);
```

Comment:
Due to some legacy reasons, the current `project_version` table might be messy. It does not correspond what you observe in `project_files` table. For instance, project `az_test` might only have one entry in `project_files`, but it has too many entries in `project_versions`. Have you considered that? Not very sure about that because we changed schema by bringing `resouce_id`, but we need to be careful about this.